src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegmentPool.java

```diff
@@ -0,0 +1,298 @@
+  }
+
+  /**
+   * Constructs new pool with specified maximum number of elements per bucket.
+   *
+   * There's {@link Long#SIZE} + 1 buckets, while last bucket is not used,
+   * as allocations with size 0 or with highest bit set to 1 goes there.
+   *
+   * @param maxSizes the array of maximum sizes per segment bucket
+   * @param scope the scope to which this allocator should be bound
+   */
+  public MemorySegmentPool(int maxSizes[], ResourceScope scope) {
+    this.scope = scope;
+
+    validateMaxSizes(maxSizes);
+    for (int i=0; i < segmentsDequeue.length; i++) {
+      var segmentsBucketMaxSize = i < maxSizes.length ? maxSizes[i] : 0;
+      segmentsDequeue[i] = new SpinLockQueue<>(segmentsBucketMaxSize);
+    }
+
+    scope.addOnClose(this::freePool);
+  }
+
+  public SegmentAllocator allocatorForScope(ResourceScope resourceScope) {
+    // Prevent scope managing this pool to go away, when dependant allocator is alive
+    final var handle = scope.acquire();
+    resourceScope.addOnClose(handle::close);
+    return (bytesSize, bytesAlignment) -> getAsNewSegmentWithScope(resourceScope, bytesSize, bytesAlignment);
+  }
+
+  @ForceInline
+  public MemoryPoolSegment getSegmentEntryByLayout(MemoryLayout layout) {
+    return getSegmentEntryBySize(layout.byteSize(), layout.byteAlignment());
+  }
+
+  /**
+   * Gets segment from pool or allocates new one. Internally segments are cached.
+   * The size of segment can be larger than requested.
+   *
+   * @param size the size of segment.
+   *
+   * @return segment of size at least `size`
+   */
+//  @ForceInline
+  public MemoryPoolSegment getSegmentEntryBySize(long size, long alignment) {
+//    if (!scope.isAlive()) {
+//      throw new IllegalStateException("Associated resource scope is closed");
+//    }
+
+    final var bitBound = bitBound(size, alignment);
+    MemoryPoolSegment segment = getMemoryPoolSegment(bitBound);
+
+    return segment;
+  }
+
+  @ForceInline
+  private MemoryPoolSegment getMemoryPoolSegment(int bitBound) {
+    final var segmentDequeue = segmentsDequeue[bitBound];
+
+    var segment = segmentDequeue.pollEntry();
+    if (segment == null) {
+      final var bitBoundedSize = 1L << bitBound;
+      segment = allocateNewEntry(segmentDequeue, bitBoundedSize);
+//      segment.memorySegment = (NativeMemorySegmentImpl) segment.memoryAddress.asSegment(bitBoundedSize, scope);
+//      segment.size = bitBoundedSize;
+    }
+    return segment;
+  }
+
+  @ForceInline
+  private MemorySegment getAsNewSegmentWithScope(ResourceScope resourceScope, long size, long alignment) {
+    final var bitBound = bitBound(size, alignment);
+    final var segmentEntry = getMemoryPoolSegment(bitBound);
+
+    ((ResourceScopeImpl) resourceScope).addOrCleanupIfFail(new ResourceCleanup() {
+      @Override
+      public void cleanup() {
+        segmentEntry.close();
+      }
+    });
+
+    return segmentEntry.memoryAddress.asSegment(1L << bitBound, resourceScope);
```

Comment:
I wouldn't be surprised if this was the issue - the segment is created with a size that is determined by a long shift. If this segment gets scalarized, I wonder if the VM is able to see that the segment size is actually an int and not a long. Worth a shot.

Reply:
I checked it, and I get same results. I'll shortly send an updated version

Reply:
Catched! It's not int, I use calculate leading zeros, I guess this a barrier for VM. Replacing size by size with align makes it fast.

But yes it's about size.

Reply:
Good catch - I'm not sure I follow the various @DontInline and such - I hope the code is not too tailored to the specific benchmark. But the numbers are rather good - now even the clean API gives numbers that beat malloc considerably (I honestly would leave the direct API alone). This is indeed a good direction - while I'm not planning to add something like this at this stage, we will, at some point after 17, ramp up some discussion around allocation strategy, and it seems like we got a good contender here ;-)

Reply:
I should thank you for help and motivation.

Now I've got highly tuned allocator to implement native I/O.

That's will be critical to provide buffers between native and heap memory.

Thanks again!